src/platform/permissions.ts

```diff
@@ -206,168 +190,162 @@ export function createAll(
+						({ permission }) => permission.__id,
 					);
-					const addPermissionsPromise = apiTx
-						.get({
-							resource: 'role__has__permission',
-							options: {
-								$select: 'permission',
-								$filter: {
-									role: role.id,
-									permission: { $in: permissions },
-								},
-							},
-						})
-						.then((rolePermissions: AnyObject[]) => {
-							const rolePermissionIds: number[] = rolePermissions.map(
-								({ permission }) => permission.__id,
-							);
-							return _.difference(permissions, rolePermissionIds);
-						})
-						.map(permission =>
-							apiTx.post({
-								resource: 'role__has__permission',
-								body: {
-									role: role.id,
-									permission,
-								},
-								options: { returnResource: false },
-							}),
-						);
-					const deletePermissionsPromise = apiTx.delete({
+					return _.difference(permissions, rolePermissionIds);
+				})
+				.map(permission =>
+					apiTx.post({
 						resource: 'role__has__permission',
-						options: {
-							$filter: {
-								role: role.id,
-								$not: { permission: { $in: permissions } },
-							},
+						body: {
+							role: role.id,
+							permission,
 						},
-					});
-					return Promise.all([addPermissionsPromise, deletePermissionsPromise]);
-				});
-			})
-			.tapCatch(err => {
-				captureException(err, `Error on configuring ${roleName}`);
+						options: { returnResource: false },
+					}),
+				);
+			const deletePermissionsPromise = apiTx.delete({
+				resource: 'role__has__permission',
+				options: {
+					$filter: {
+						role: role.id,
+						$not: { permission: { $in: permissions } },
+					},
+				},
 			});
+			await Promise.all([addPermissionsPromise, deletePermissionsPromise]);
+			return role;
+		} catch (err) {
+			captureException(err, `Error on configuring ${roleName}`);
+			throw err;
+		}
+	};
 
-	const rolesPromise = Promise.props<Dictionary<{ id: number }>>(
+	const rolesPromise = Bluebird.props<Dictionary<{ id: number }>>(
 		_.mapValues(roleMap, createRolePermissions),
-	)
-		.tap(roles =>
-			// Assign user roles
-			Promise.all(
-				_.map(userMap, (userEmails, roleName) =>
-					Promise.mapSeries(userEmails, email =>
-						findUser(email, tx)
-							.then(user => {
-								if (user == null || user.id == null) {
-									throw new Error(`User ${email} not found.`);
-								}
-								return assignUserRole(user.id, roles[roleName].id, tx);
-							})
-							.catch(_.noop),
-					),
-				),
-			),
-		)
-		.tap(() =>
-			// Remove stale permissions, preserving unassigned ones.
-			permissionsCache.then(permissions =>
-				apiTx
-					.delete({
-						resource: 'permission',
-						options: {
-							$filter: {
-								$not: {
-									$or: [
-										{
-											is_of__role: {
-												$any: {
-													$alias: 'rhp',
-													$expr: { rhp: { id: { $ne: null } } },
-												},
-											},
+	).tap(async roles => {
+		// Assign user roles
+		await Bluebird.all(
```

Comment:
```suggestion
		await Bluebird.map(_.toPairs(userMap), async ([userEmails, roleName]) => {
```

Reply:
The current form will be quicker because it avoids creating unnecessarily an unnecessary intermediate